### ISSUE TITLE
Training cannot be continued with --regcb or --regcbopt after model reload (--save_resume turned on).

### ISSUE THREAD

**Issue by alxbar75** [VowpalWabbit/vowpal_wabbit#3179]
#### Describe the bug
A vowpal wabbit model, initialized with --regcb or --regcbopt and --save_resume, is not improving during training after it has been reloaded.

#### To Reproduce
Example Code, 6 different contexts with 3 different actions, each combination having a probability of giving a reward (a click, cost=-1). Without saving/loading, the model converges to a good and sometimes optimal solution. With save/loading, the model is behaving differently.

To see convergence set `store_and_load = False`
```python
import operator
import random
import numpy as np
from matplotlib import pyplot as plt
from vowpalwabbit import pyvw

plt.ion()

def running_mean(x, N):
    cumsum = np.cumsum(np.insert(x, 0, 0))
    return (cumsum[N:] - cumsum[:-N]) / float(N)


def to_vw_example_format(context, cb_label=None):
    if cb_label is not None:
        chosen_action, cost, prob = cb_label
    example_string = ""
    example_string += "shared |User device={} \n".format(context)
    for action in action_space:
        if cb_label is not None and action == chosen_action:
            example_string += "1:{}:{} ".format(cost, prob)
        example_string += "|Action ad={} \n".format(action)
    # Strip the last newline
    return example_string[:-1]


# definition of problem to solve. Goal: playing out the article with highest ctr given a device
context_to_action_ctr = {
    "device-1": {"article-1": 0.05, "article-2": 0.06, "article-3": 0.04},  # 2
    "device-2": {"article-1": 0.08, "article-2": 0.07, "article-3": 0.05},  # 1
    "device-3": {"article-1": 0.01, "article-2": 0.04, "article-3": 0.09},  # 3
    "device-4": {"article-1": 0.04, "article-2": 0.04, "article-3": 0.045},  # 3
    "device-5": {"article-1": 0.09, "article-2": 0.01, "article-3": 0.07},  # 1
    "device-6": {"article-1": 0.03, "article-2": 0.09, "article-3": 0.04}  # 2
}
action_space = list(context_to_action_ctr["device-1"].keys())

vw = f"--cb_explore_adf -q UA -q UU --save_resume --regcbopt --mellowness 0.0003 --power_t 0 -l 0.5"

actor = pyvw.vw(vw)

# vw performs better with this variable set to False
store_and_load = False

random_rewards = []
actor_rewards = []
optimal_rewards = []

for step in range(2000000):

    # pick a random context
    device = random.choice(list(context_to_action_ctr.keys()))

    # let vw generate probability distribution
    action_probabilities = np.array(actor.predict(to_vw_example_format(device)))

    # sample action
    probabilities = action_probabilities / action_probabilities.sum()
    action_idx = np.random.choice(len(probabilities), 1, p=probabilities)[0]
    probability = action_probabilities[action_idx]

    # get reward/regret
    action_to_reward_regret = {
        action: (1, 0) if random.random() < context_to_action_ctr[device][action] else (0, 1) for action in action_space
    }

    actor_action = action_space[action_idx]
    random_action = random.choice(action_space)
    optimal_action = max(context_to_action_ctr[device].items(), key=operator.itemgetter(1))[0]

    # update statistics
    actor_rewards.append(action_to_reward_regret[actor_action][0])
    random_rewards.append(action_to_reward_regret[random_action][0])
    optimal_rewards.append(action_to_reward_regret[optimal_action][0])

    # learn online
    reward, regret = action_to_reward_regret[actor_action]
    cost = -1 if reward == 1 else 0
    actor.learn(to_vw_example_format(device, (actor_action, cost, probability)))

    if step % 1000 == 0 and step >= 5000:
        plt.clf()
        axes = plt.gca()
        plt.title("Reward over time")
        plt.plot(running_mean(actor_rewards, 10000), label=str(vw))
        plt.plot(running_mean(random_rewards, 10000), label="Random actions")
        plt.plot(running_mean(optimal_rewards, 10000), label="Optimal actions")
        plt.legend()
        plt.pause(0.0001)

        print("POLICY")
        for device in list(context_to_action_ctr.keys()):
            prediction = np.array(actor.predict(to_vw_example_format(device)))
            print(device, list("%.2f" % pred for pred in prediction), np.argmax(prediction) + 1)


        if store_and_load:
            actor.save("model.vw")
            vw = pyvw.vw("--quiet -i model.vw")
            actor = vw

```

#### Expected behavior
The model should behave exactly the same after reloading, training should lead to convergence after save/load.

#### Observed Behavior
The model does not converge after reloading.

#### Environment
vowpalwabbit==8.10.2
python 3.8.5
Ubuntu 20.04

#### Additional context
We would like to have a scalable system with training and model-inference separated. The model is trained periodically on new data.

**Comment by ataymano** [comment]
hi,
thank you for reporting this.
Confirmed that:
1) Issue is reproducible with python. Small clarification: problem is not about resetting state every time we save the model, but with inability to learn at all after single save-load cycle. I updated largest probabilities to 0.9 for each device to make effect more visible and have that kind of plot:
![image](https://user-images.githubusercontent.com/41013086/126637865-d564735e-53e4-42c6-97bc-8ac86edc6366.png)
2) Not able to reproduce the same issue with vw command line

**Comment by alxbar75** [comment]
I experience a similar problem with --bag
I can't say if training completely halts but convergence without save/load is definitely better.

```
import operator
import random
import numpy as np
from matplotlib import pyplot as plt
from vowpalwabbit import pyvw

plt.ion()

def running_mean(x, N):
    cumsum = np.cumsum(np.insert(x, 0, 0))
    return (cumsum[N:] - cumsum[:-N]) / float(N)


def to_vw_example_format(context, cb_label=None):
    if cb_label is not None:
        chosen_action, cost, prob = cb_label
    example_string = ""
    example_string += "shared |User device={} \n".format(context)
    for action in action_space:
        if cb_label is not None and action == chosen_action:
            example_string += "1:{}:{} ".format(cost, prob)
        example_string += "|Action ad={} \n".format(action)
    # Strip the last newline
    return example_string[:-1]


# definition of problem to solve. Goal: playing out the article with highest ctr given a device
context_to_action_ctr = {
    "device-1": {"article-1": 0.05, "article-2": 0.06, "article-3": 0.04},  # 2
    "device-2": {"article-1": 0.08, "article-2": 0.07, "article-3": 0.05},  # 1
    "device-3": {"article-1": 0.01, "article-2": 0.04, "article-3": 0.09},  # 3
    "device-4": {"article-1": 0.04, "article-2": 0.04, "article-3": 0.045},  # 3
    "device-5": {"article-1": 0.09, "article-2": 0.01, "article-3": 0.07},  # 1
    "device-6": {"article-1": 0.03, "article-2": 0.09, "article-3": 0.04}  # 2
}
action_space = list(context_to_action_ctr["device-1"].keys())

vw = f"--cb_explore_adf -q UA -q UU --quiet --save_resume --bag 10 --epsilon 0.1"

actor = pyvw.vw(vw)

# vw performs better with this variable set to False
store_and_load = True

train_freq = 100

random_rewards = []
actor_rewards = []
optimal_rewards = []
training_instances = []
for step in range(60000):

    # pick a random context
    device = random.choice(list(context_to_action_ctr.keys()))

    # let vw generate probability distribution
    action_probabilities = np.array(actor.predict(to_vw_example_format(device)))

    # sample action
    probabilities = action_probabilities / action_probabilities.sum()
    action_idx = np.random.choice(len(probabilities), 1, p=probabilities)[0]
    probability = action_probabilities[action_idx]

    # get reward/regret
    action_to_reward_regret = {
        action: (1, 0, context_to_action_ctr[device][action]) if random.random() < context_to_action_ctr[device][action] else (0, 1, context_to_action_ctr[device][action]) for action in action_space
    }

    actor_action = action_space[action_idx]
    random_action = random.choice(action_space)
    optimal_action = max(context_to_action_ctr[device].items(), key=operator.itemgetter(1))[0]

    # update statistics
    actor_rewards.append(action_to_reward_regret[actor_action][2])
    random_rewards.append(action_to_reward_regret[random_action][2])
    optimal_rewards.append(action_to_reward_regret[optimal_action][2])

    # learn online
    reward, regret, ctr = action_to_reward_regret[actor_action]
    cost = -1 if reward == 1 else 0
    # actor.learn(to_vw_example_format(device, (actor_action, cost, probability)))

    training_instances.append((actor_action, cost, device))

    if step % train_freq == 0:

        # random.shuffle(training_instances)
        # with tqdm.tqdm(total=len(training_instances)) as pbar:
        for actor_action, cost, device in training_instances:
            actor.learn(to_vw_example_format(device, (actor_action, cost, 1.0)))

        training_instances = []

        if store_and_load:
            actor.save("model.vw")
            vw = pyvw.vw("--quiet -i model.vw")
            actor = vw


    if step % 1000 == 0 and step >= 5000:
        plt.clf()
        axes = plt.gca()
        plt.title("Reward over time")
        plt.plot(running_mean(actor_rewards, 100), label=str(vw))
        plt.plot(running_mean(random_rewards, 100), label="Random actions")
        plt.plot(running_mean(optimal_rewards, 100), label="Optimal actions")
        plt.legend()
        plt.pause(0.0001)

        print("POLICY")
        for device in list(context_to_action_ctr.keys()):
            prediction = np.array(actor.predict(to_vw_example_format(device)))
            print(device, list("%.2f" % pred for pred in prediction), np.argmax(prediction) + 1)

plt.ioff()
plt.show()
```
With save/load:

![image](https://user-images.githubusercontent.com/61507801/126676977-67d5fa8c-92b7-4ce0-93ce-78c5e1fd48e0.png)

Without save/load:

![image](https://user-images.githubusercontent.com/61507801/126677078-5ec3de10-13a4-42af-a4ac-8883523d8008.png)

**Comment by ataymano** [comment]
hmm, that one could be different.
regcb one can be easily reproduced with making problem easier by bumping best arms to 0.9, while this one cannot.
So either there is some another (and more subtle) bug here, or it is just slower convergence because of bad luck. 0.05 difference in performance is something "we don't know right answers for couple of contexts yet", and given how ambiguous is the problem, time for us to learn all 6 contexts could vary. Let me double check, but is this effect reproducible for you?

**Comment by alxbar75** [comment]
Hey, yes I think this problem is somewhat different and more sublte. 
I have been starting that script multiple times and I experienced the same problem:
With save/load: definetly better than random but not as good and with a more unstable policy
Without save/load: approaching optimal more often and staying at a higher level (expected ctr) while still exploring

However this setup is also somewhat special, as training data is collected and trained on batch-wise, rather than online per instance. (However, it still converges fairly good)

**Comment by alxbar75** [comment]
I think I found the reason for the difference in performance. I did't use --save_resume when loading the model, just during initialization. Could this be the reason? Do I also have to use --save_resume when loading a model?

**Comment by ataymano** [comment]
hi,
1) you don't have to use save_resume when loading the model - it is set automatically if model was saved with --save_resume
2) when you see improvement, which scenario are you talking about - regcb or bagging?
- For regcb I was not able to observe an improvement after adding --save_resume (as expected actually). Will look closer where is the problem.
- for --bagging I am still struggling to reproduce original problem: I somehow have proper learning plots when I am running your code. Can you please show your 'pip list | grep vowpalwabbit' output?

**Comment by alxbar75** [comment]
Hi,

`vowpalwabbit             8.10.2`

I was referring to --bagging. If you take the second script I posted and add "--save_resume" to line 96
resulting in `vw = pyvw.vw("--quiet -i model.vw --save_resume")` It seems to me that performance is increasing.


With --save_resume during model loading:
![with_save_resume](https://user-images.githubusercontent.com/61507801/127515766-0d1c9ae2-7093-43d4-9245-7af350b9347b.png)

Without --save_resume during model loading
(but still during initialization vw = f"--cb_explore_adf -q UA -q UU --quiet --save_resume --bag 10 --epsilon 0.1")
![without_save_resume](https://user-images.githubusercontent.com/61507801/127515988-e5e90a1a-1517-4fd2-bb80-669e5a9a8d7f.png)

Starting the script multiple times, leads in both cases to similar results

**Comment by alxbar75** [comment]
Just checked, same behavior with `vowpalwabbit             8.11.0`

**Comment by ataymano** [comment]
Sorry for delay in response.
Not sure about all details, but looks like root cause is with messed probabilities with regcb case.

1) Command line experiment:
On the following dataset:
```
shared |User device=1 
0:-1:0.5 |Action A 
|Action B 

shared |User device=1 
0:-1:0.5 |Action A 
|Action B 

```
predictions from the following command line:
`vw --cb_explore_adf -q UA -q UU --save_resume --regcbopt --mellowness 0.0003 --power_t 0 -l 0.5 -d data.txt -p pred.txt`
looks like this:
```
0:0,1:0

0:-0.393469,1:-0.236082
```
2) "equivalent" pyvw experiment:
```
from vowpalwabbit import pyvw
agent = pyvw.vw(f"--cb_explore_adf -q UA -q UU --save_resume --regcbopt --mellowness 0.0003 --power_t 0 -l 0.5")

device = 'device-1'

example = 'shared |User device=1 \n' + '0:-100:0.5 |Action A\n' + '|Action B'

print(f'Prediction before learning = {agent.predict(example)}')
agent.learn(example)
print(f'Prediction after learning = {agent.predict(example)}')
```

Output:
```
Prediction before learning = [1.0, 1.0]
Prediction after learning = [1.0, 1.0]
```

So, looks like the problem is with negative probabilities which are somewhat reasonably handled by vw binary, but got messed with normalization logic in python (which can lead either to uniform distribution if all values are truncated to 0 or 1).
Looking deeper.

**Comment by bassmang** [comment]
Hi @alxbar75,

Apologies for the delay with this issue. I believe the issue here is actually that `power_t` and `learning_rate` are not being persisted when a model is saved and loaded. Consider this snippet which is similar to your example but sets the `random` seed and `np.random` seed to 0 for consistent randomness and removes the plotting. 

```
import operator
import random
import numpy as np
from matplotlib import pyplot as plt
from vowpalwabbit import pyvw

STORE_AND_LOAD = True
random.seed(0)
np.random.seed(0)

def running_mean(x, N):
    cumsum = np.cumsum(np.insert(x, 0, 0))
    return (cumsum[N:] - cumsum[:-N]) / float(N)


def to_vw_example_format(context, cb_label=None):
    if cb_label is not None:
        chosen_action, cost, prob = cb_label
    example_string = ""
    example_string += "shared |User device={} \n".format(context)
    for action in action_space:
        if cb_label is not None and action == chosen_action:
            example_string += "1:{}:{} ".format(cost, prob)
        example_string += "|Action ad={} \n".format(action)
    # Strip the last newline
    return example_string[:-1]


# definition of problem to solve. Goal: playing out the article with highest ctr given a device
context_to_action_ctr = {
    "device-1": {"article-1": 0.05, "article-2": 0.06, "article-3": 0.04},  # 2
    "device-2": {"article-1": 0.08, "article-2": 0.07, "article-3": 0.05},  # 1
    "device-3": {"article-1": 0.01, "article-2": 0.04, "article-3": 0.09},  # 3
    "device-4": {"article-1": 0.04, "article-2": 0.04, "article-3": 0.045},  # 3
    "device-5": {"article-1": 0.09, "article-2": 0.01, "article-3": 0.07},  # 1
    "device-6": {"article-1": 0.03, "article-2": 0.09, "article-3": 0.04}  # 2
}
action_space = list(context_to_action_ctr["device-1"].keys())

vw = f"--cb_explore_adf -q UA -q UU --save_resume --regcbopt --mellowness 0.0003 --power_t 0 -l 0.5"

actor = pyvw.vw(vw)

random_rewards = []
actor_rewards = []
optimal_rewards = []

for step in range(20000):

    # pick a random context
    device = random.choice(list(context_to_action_ctr.keys()))

    # let vw generate probability distribution
    action_probabilities = np.array(actor.predict(to_vw_example_format(device)))

    # sample action
    probabilities = action_probabilities / action_probabilities.sum()
    action_idx = np.random.choice(len(probabilities), 1, p=probabilities)[0]
    probability = action_probabilities[action_idx]

    # get reward/regret
    action_to_reward_regret = {
        action: (1, 0) if random.random() < context_to_action_ctr[device][action] else (0, 1) for action in action_space
    }

    actor_action = action_space[action_idx]
    random_action = random.choice(action_space)
    optimal_action = max(context_to_action_ctr[device].items(), key=operator.itemgetter(1))[0]

    # update statistics
    actor_rewards.append(action_to_reward_regret[actor_action][0])
    random_rewards.append(action_to_reward_regret[random_action][0])
    optimal_rewards.append(action_to_reward_regret[optimal_action][0])

    # learn online
    reward, regret = action_to_reward_regret[actor_action]
    cost = -1 if reward == 1 else 0
    actor.learn(to_vw_example_format(device, (actor_action, cost, probability)))

    if step % 1000 == 0 and step >= 5000:

        if STORE_AND_LOAD:
            actor.save("model.vw")
            vw = pyvw.vw("--quiet -i model.vw --power_t 0 -l 0.5")
            actor = vw

print(running_mean(actor_rewards, 10000)[-1])
```

The final output here will be `0.0669` here regardless of `STORE_AND_LOAD`. I have set `--power_t 0 -l 0.5` in the loaded model to be consistant with the original model. It does seem that these values should be persisted for a loaded model so I've put in a PR to keep these and other `update` arguments.

**Comment by bassmang** [comment]
Predictions were being altered in learn -- need to store before learning, and move after.
